Blog post: introducing KW 1.11 (#217)
kubewarden/kubewarden.io: content/blog/2024/03/kubewarden-1-11-release.md

---
title: Kubewarden 1.11 release
authors:
  - Flavio Castelli
date: 2024-03-14
---

Today we're glad to announce the release of Kubewarden 1.11.

> This release focused on performance improvements, especially when running on big Kubernetes clusters.

Reviewer: focuses

Author: done

## Audit Scanner

A lot of work has been done on the audit scanner. The auditing of resource is now done in parallel, which means less time is required on big clusters to scan all the available resources.

We've also changed how we handle Policy Reports. Kubewarden is still using the Policy Report format being defined inside of the Kubernetes's `wg-policy` group. However, prior to this release,
we used to create one `PolicyReport` per `Namespace` and one `ClusterPolicyReport` per cluster.

> This solution proved to not scale inside of clusters with many audited resources. That's why, starting from Kubewarden 1.11, the audit scanner

Author: I've expanded on that, please take another look

---
title: Kubewarden 1.11 release
authors:
  - Flavio Castelli
date: 2024-03-14
---

Today we're glad to announce the release of Kubewarden 1.11.

This release focuses on performance improvements, especially when running on big Kubernetes clusters.

## Audit Scanner

A lot of work has been done on the audit scanner. The auditing of resource is now done in parallel, which means less time is required on big clusters to scan all the available resources.

> We've also changed how we handle Policy Reports. Kubewarden is still using the Policy Report format being defined inside of the Kubernetes's `wg-policy` group. However, prior to this release,
> we used to create one `PolicyReport` per `Namespace` and one `ClusterPolicyReport` per cluster.
> This solution proved to not scale inside of clusters with many audited resources. Depending on the number of policies and audited resources, the size of the final report objects could be too big to fit into etcd.
> Moreover, the amount of memory required by the audit scanner could be significant.

Reviewer 2: ```suggestion
We've also changed how we handle Policy Reports. Kubewarden is still using the Policy Report format being defined inside the Kubernetes's `wg-policy` group. However, prior to this release,
we used to create one `PolicyReport` per `Namespace` and one `ClusterPolicyReport` per cluster.
This solution proved to not scale inside clusters with many audited resources. Depending on the number of policies and audited resources, the size of the final report objects could be too big to fit into etcd.
Moreover, the amount of memory required by the audit scanner could be significant.
```

> That's why, starting from Kubewarden 1.11, the audit scanner creates one `PolicyReport` per namespaced resource and one `ClusterPolicyReport` per cluster-wide resource being scanned.
> That ensures the reports can always fit etcd's size requirement. On top of that, now the audit scanner process consumes always the same amount of memory, regardless of the number of
> policies and resources being audited.

Reviewer 2: ```suggestion
That's why, starting from Kubewarden 1.11, the audit scanner creates one `PolicyReport` per namespaced resource and one `ClusterPolicyReport` per cluster-wide resource being scanned.
That ensures the reports can always fit etcd's size requirement. On top of that, now the audit scanner process always consumes the same amount of memory, regardless of the number of
policies and resources being audited.
```

> Getting the auditing results of the resources defined inside of a namespace can be done in this way:

Reviewer 2: ```suggestion
Getting the auditing results of the resources defined inside a namespace can be done in this way:
```

04937b2b-e68b-47d5-909d-d0ae75527f07   Pod          pod2                        9      1      0      0       0      1h
...

```

While the auditing results of cluster-wide resources can be obtained in this way:

```console
kubectl get cpolr -o wide

NAME                                   KIND        NAME                 PASS   FAIL   WARN   ERROR   SKIP   AGE
261c9492-deec-4a09-8aa9-cd464bb4b8d1   Namespace   namespace1           3      1     0       0       0      1h
35ca342f-685b-4162-a342-8d7a52a61749   Namespace   namespace2           0      4     0       0       0      1h
3a8f8a88-338b-4905-b9e4-f13397a0d7b5   Namespace   namespace3           4      0     0       0       0      15h
```

Take a look at [the official documentation](https://docs.kubewarden.io/explanations/audit-scanner/policy-reports) to learn more about policy reports.

## New host capability available

Kubewarden provides a set of [host capabilities](https://docs.kubewarden.io/reference/spec/host-capabilities/intro-host-capabilities) that allow
policy authors to access external data during policy evaluation.

> Kubewarden 1.11 expands the operations that can be done against a container registry by making possible to fetch OCI manifests.
> This addition was requested by our community as a way to write policies that make use of the platform and architecture information found inside of the OCI manifest to make mutations.

Reviewer 2: ```suggestion
Kubewarden 1.11 expands the operations that can be done against a container registry by making it possible to fetch OCI manifests.
This addition was requested by our community as a way to write policies that make use of the platform and architecture information found inside the OCI manifest to make mutations.
```

The Kubewarden SDKs have been updated to expose this new functionality. We will also publish a dedicated blog post explaining what policy authors can achieve with this new capability.

## Context aware policies optimization

As part of our host capabilities, Kubewarden provides a way to query the Kubernetes API server. This is used to create [context aware policies](https://docs.kubewarden.io/next/reference/spec/context-aware-policies).
These are policies that use information about the state of Kubernetes cluster to make validating/mutating decisions.

> Starting from the 1.11 release, Kubewarden uses a different way to fetch information from the Kubernetes API server and keep this data up to date. That brings to major improvements.
> First, the load on the Kubernetes API server is reduced. Second, the information shared with the policies is significantly fresher than before. That means changes done inside of Kubernetes are propagated faster to

Reviewer 2: ```suggestion
Starting from the 1.11 release, Kubewarden uses a different way to fetch information from the Kubernetes API server and keep this data up to date. That brings two major improvements.
Firstly, the load on the Kubernetes API server is reduced. Secondly, the information shared with the policies is significantly fresher than before. That means changes done inside Kubernetes are propagated faster to
```

the policies.

## Reduce latency of policies that make use of host capabilities

Thanks to a report coming from our community, we learnt that policies making use of host capabilities had too much latency when ran on big clusters.

> We started an [investigation](https://github.com/kubewarden/policy-server/issues/692) that lead to several performance improvements that are now part of the Kubewarden 1.11 release.

Reviewer 2: ```suggestion
We started an [investigation](https://github.com/kubewarden/policy-server/issues/692) that led to several performance improvements that are now part of the Kubewarden 1.11 release.
```
